File: examples/good/typescript/postsRouter.ts
import express from 'express';
import { PostService } from './services/PostService';

export function createPostsRouter(postService: PostService): express.Router {
  const router = express.Router();

  router.get('/posts', async (req, res, next) => {
    try {
      const limit = parseInt(req.query.limit as string, 10) || 20;
      const cursor = (req.query.cursor as string) || undefined;

      const posts = await postService.listPosts(limit, cursor);

      res.status(200).json({
        data: posts,
        pagination: {
          next_cursor: null,
          previous_cursor: null,
          limit,
        },
      });
    } catch (err) {
      next(err);
    }
  });

  router.post('/posts', async (req, res, next) => {
    try {
      const { title, body, author_id } = req.body ?? {};

      if (!title || !body || !author_id) {
        return res.status(422).json({
          error: {
            code: 'VALIDATION_ERROR',
            message: 'Missing required fields.',
            details: {
              title: title ? [] : ['This field is required.'],
              body: body ? [] : ['This field is required.'],
              author_id: author_id ? [] : ['This field is required.'],
            },
          },
        });
      }

      const post = await postService.createPost({
        title,
        body,
        authorId: author_id,
      });

      res.status(201).json(post);
    } catch (err) {
      if (err instanceof Error) {
        return res.status(422).json({
          error: {
            code: 'VALIDATION_ERROR',
            message: err.message,
          },
        });
      }
      next(err);
    }
  });

  return router;
}

File: examples/good/python/posts_router.py
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, status

from .dto import Post, PostCreate
from .repositories import InMemoryPostRepository
from .services import PostService

router = APIRouter(prefix="/posts", tags=["posts"])

_repo = InMemoryPostRepository()
_service = PostService(_repo)


def get_post_service() -> PostService:
    return _service


@router.get("", response_model=dict)
def list_posts(
    limit: int = 20,
    cursor: Optional[str] = None,
    service: PostService = Depends(get_post_service),
):
    posts = service.list_posts(limit, cursor)
    return {
        "data": [p.dict() for p in posts],
        "pagination": {
            "next_cursor": None,
            "previous_cursor": None,
            "limit": limit,
        },
    }


@router.post("", response_model=Post, status_code=status.HTTP_201_CREATED)
def create_post(
    payload: PostCreate,
    service: PostService = Depends(get_post_service),
):
    # In a real app, author_id comes from auth dependency.
    author_id = "user_1"
    try:
        return service.create_post(payload, author_id)
    except ValueError as exc:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail=str(exc),
        ) from exc
